### PR TITLE
[k8s] Fix mounting when launching from a service account

### LINKED ISSUE #3530: [k8s] file mount fails when using job controller
I had the following task and config which I executed on k8s with `sky jobs launch -n job5 sky-kueue-task.yaml`

` sky-kueue-task.yaml`
```
resources:
  cloud: kubernetes
  accelerators: H100:8
  labels:
    kueue.x-k8s.io/queue-name: user-queue-a

num_nodes: 2

file_mounts:
  ~/spin.py: ./spin.py

setup: |
  pip install torch

run: |
  python ~/spin.py
```
`~/.sky/config.yaml`
```
kubernetes:
  provision_timeout: -1  # Wait indefinitely in the queue to get scheduled on the k8s cluster
  remote_identity: SERVICE_ACCOUNT
```
This hangs indefinitely. When I log into the jobs controller, it looks like there's a 403 error on `skypilot-service-account` for creating a namespace during the FUSE device manager creation.

```
I 05-09 18:36:18 config.py:455] _configure_fuse_mounting: Setting up FUSE device manager.
E 05-09 18:36:18 provisioner.py:91] Failed to configure 'job5-2' on Kubernetes Region(name='kubernetes') (all zones) with the following error:
E 05-09 18:36:18 provisioner.py:91] kubernetes.client.exceptions.ApiException: (403)
E 05-09 18:36:18 provisioner.py:91] Reason: Forbidden
E 05-09 18:36:18 provisioner.py:91] HTTP response headers: HTTPHeaderDict({'Audit-Id': '994b9637-1caa-4fbd-ba89-b27de4fcfcd3', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': 'aef310ce-b0dc-4769-9e55-1a2febf4a813', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'c67a67d7-3d69-4dea-90b4-7b626853112c', 'Date': 'Thu, 09 May 2024 18:36:18 GMT', 'Content-Length': '311'})
E 05-09 18:36:18 provisioner.py:91] HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"namespaces is forbidden: User \"system:serviceaccount:default:skypilot-service-account\" cannot create resource \"namespaces\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"namespaces"},"code":403}
```

_Version & Commit info:_
* `sky -v`: 1.0.0.dev20240509
* `sky -c`: 07da93b3c3b87e53d0ebd7e033c9ce24992151fe

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Closes #3530.

Adds the appropriate permissions to create the skypilot-system namespace and create resources in it.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Smoke tests - enabled `test_managed_jobs_storage` for kubernetes `pytest -v tests/test_smoke.py::test_managed_jobs_storage --kubernetes`

### CODE REVIEW

File: sky/provision/kubernetes/config.py [hunk 1]
```diff
@@ -239,7 +256,9 @@ def _configure_autoscaler_service_account(
     if len(accounts) > 0:
         assert len(accounts) == 1
         logger.info('_configure_autoscaler_service_account: '
-                    f'{using_existing_msg(account_field, name)}')
+                    f'{updating_existing_msg(account_field, name)}')
+        kubernetes.core_api().patch_namespaced_service_account(
+            name, namespace, account)
```
Comment (Michaelvll) [inline sky/provision/kubernetes/config.py:261]: Is it possible to check the config of the account first and only patch the service account if it does not have all the config?
Reason: previously, we only required read permission for the user's account, but now we require write permission as well. This is to make sure the following works: the admin create the service account for all users and create restrictive user accounts with only read permission (this is what we made sure for the other clouds). Wdyt?

File: sky/provision/kubernetes/config.py [hunk 2]
```diff
@@ -277,7 +296,8 @@ def _configure_autoscaler_role(namespace: str, provider_config: Dict[str, Any],
     if len(accounts) > 0:
         assert len(accounts) == 1
         logger.info('_configure_autoscaler_role: '
-                    f'{using_existing_msg(role_field, name)}')
+                    f'{updating_existing_msg(role_field, name)}')
+        kubernetes.auth_api().patch_namespaced_role(name, namespace, role)
```
Comment (Michaelvll) [inline sky/provision/kubernetes/config.py:300]: Similar as above : )

File: sky/provision/kubernetes/config.py [hunk 3]
```diff
@@ -438,6 +471,48 @@ def _configure_ssh_jump(namespace, config: common.ProvisionConfig):
...
+    """
+    svc_account_namespace = provider_config['namespace']
+    skypilot_system_namespace = provider_config['skypilot_system_namespace']
+    kubernetes_utils.create_namespace(skypilot_system_namespace)
+
+    # Setup permissions if using the default service account.
+    # If the user has requested a different service account (via
+    # remote_identity in ~/.sky/config.yaml), we assume they have already set
+    # up the necessary roles and role bindings.
+    if service_account == 'skypilot-service-account':
```
Comment (Michaelvll) [inline sky/provision/kubernetes/config.py:490]: Should we use a constant for this `skypilot-service-account`?

File: sky/provision/kubernetes/config.py [hunk 4]
```diff
@@ -446,17 +521,15 @@ def _configure_fuse_mounting(provider_config: Dict[str, Any]) -> None:
...
     SkyPilot pod requests this resource to mount the FUSE filesystem.
 
-    We create this daemonset in a common namespace, which is configurable in the
-    provider config. This allows the FUSE mounting sidecar to be shared across
-    multiple tenants. The default namespace is 'sky-system' (populated in
-    clouds.Kubernetes)
+    We create this daemonset in the skypilot_system_namespace, which is
+    configurable in the provider config. This allows the FUSE mounting sidecar
+    to be shared across multiple tenants. The default namespace is
+    'sky-system' (populated in clouds.Kubernetes).
```
Comment (Michaelvll) [inline sky/provision/kubernetes/config.py:527]: it seems we are using `skypilot-system` instead, should we update the comment here?
```suggestion
    'skypilot-system' (populated in clouds.Kubernetes).
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the fix @romilbhardwaj! Mostly looks good to me : )

File: sky/provision/kubernetes/config.py (hunk 1 above)
Comment (romilbhardwaj) [reply sky/provision/kubernetes/config.py]: Good point! I've updated checks to patch _only_ if the existing role does not match the required permissions.

**Comment by romilbhardwaj** [comment]
Thanks @Michaelvll! Running smoke tests now:

- [ ] Smoke tests - `pytest tests/test_smoke.py --kubernetes --managed-jobs`
- [ ] Smoke tests -`pytest tests/test_smoke.py --kubernetes --serve`

**Review by Michaelvll — APPROVED** [review]

**Comment by romilbhardwaj** [comment]
Tests pass, merging now!

**Pull request merged**